### PR TITLE
Experimental fast validator code path, using cwl-utils

### PR DESCRIPTION
**Pull request by tetron**
- [x] https://github.com/common-workflow-language/cwl-utils/pull/156 merged & released
- [x] documentation of limitations
- [x] tests
- [x] rename to `--fast-parser`
- [x] post-process hints to expand `class` fields in extensions

This is a work in progress.

Requires schema salad codegen improvements in https://github.com/common-workflow-language/schema_salad/pull/587 which will need to be applied to cwl-utils.

On a very large workflow I was testing with, the validation time went
120 seconds to 20 seconds.

### CODE REVIEW

**Comment by tetron** [comment]
To-do: 

[ ] fix the bugs that are breaking conformance tests (16 failures from the draft v1.2.1 tests as of this writing)
[ ] fixing type annotations

**Comment by tetron** [comment]
With parser from latest codegen fixes, now passing the entire v1.2.1-proposed conformance suite.

**Comment by mr-c** [comment]
Great progress!

> Requires schema salad codegen improvements in [common-workflow-language/schema_salad#587](https://github.com/common-workflow-language/schema_salad/pull/587) which will need to be applied to cwl-utils.

Can we get a draft PR with the new parsers?

**Comment by tetron** [comment]
> Can we get a draft `cwl-utils` PR with the new parsers?

As soon as I can make mypy happy

**Comment by tetron** [comment]
We probably also want a CI action that run the conformance tests with --fast-validator

**Comment by mr-c** [comment]
For the docs, we need to explain if there are any downsides. (Are error messages less helpful? Are some types of errors not caught by this code path?)

**Comment by tetron** [comment]
> For the docs, we need to explain if there are any downsides. (Are error messages less helpful? Are some types of errors not caught by this code path?)

There are two main gaps currently:

1. It doesn't do link checking (so things like a dangling 'source' field won't be reported, it'll fail at runtime)
2. Runtime errors won't have file/row/column at all

Validation errors do include file/row/column but are somewhat less polished.

**Comment by mr-c** [comment]
> > For the docs, we need to explain if there are any downsides. (Are error messages less helpful? Are some types of errors not caught by this code path?)
> 
> There are two main gaps currently:
> 
>     1. It doesn't do link checking (so things like a dangling 'source' field won't be reported, it'll fail at runtime)
> 
>     2. Runtime errors won't have file/row/column at all
> 
> 
> Validation errors do include file/row/column but are somewhat less polished.

Confirmed by editing `tests/echo_broken_outputs.cwl` to be a v1.2 file:

```
$ cwltool --fast-validator --validate tests/echo_broken_outputs.cwl
INFO /home/michael/cwltool/env3.10/bin/cwltool 3.1.20220902151754
INFO Resolved 'tests/echo_broken_outputs.cwl' to 'file:///home/michael/cwltool/tests/echo_broken_outputs.cwl'
ERROR Tool definition failed validation:
- tried CommandLineTool but
  Trying 'CommandLineTool'
    the `outputs` field is not valid because:
      Expected a list, was <class 'NoneType'>
- tried ExpressionTool but
  Not a ExpressionTool
- tried Workflow but
  Not a Workflow
- tried Operation but
  Not a Operation
- tried array<CommandLineTool | ExpressionTool | Workflow | Operation> but
  Expected a list, was <class 'dict'>
$ cwltool  --validate tests/echo_broken_outputs.cwl
INFO /home/michael/cwltool/env3.10/bin/cwltool 3.1.20220902151754
INFO Resolved 'tests/echo_broken_outputs.cwl' to 'file:///home/michael/cwltool/tests/echo_broken_outputs.cwl'
ERROR Tool definition failed validation:
tests/echo_broken_outputs.cwl:10:1: "outputs" section is not valid.
```
A candidate for a unit test

**Comment by mr-c** [comment]
ooh, it doesn't catch the empty `requirements` section in `tests/wc-tool-bad-reqs.cwl`
```
$ cwltool  --validate tests/wc-tool-bad-reqs.cwl
INFO /home/michael/cwltool/env3.10/bin/cwltool 3.1.20220902151754
INFO Resolved 'tests/wc-tool-bad-reqs.cwl' to 'file:///home/michael/cwltool/tests/wc-tool-bad-reqs.cwl'
ERROR Tool definition failed validation:
tests/wc-tool-bad-reqs.cwl:6:1: If 'requirements' is present then it must be a list or
                                map/dictionary, not empty.
$ cwltool --fast-validator --validate tests/wc-tool-bad-reqs.cwl
INFO /home/michael/cwltool/env3.10/bin/cwltool 3.1.20220902151754
INFO Resolved 'tests/wc-tool-bad-reqs.cwl' to 'file:///home/michael/cwltool/tests/wc-tool-bad-reqs.cwl'
tests/wc-tool-bad-reqs.cwl is valid CWL.
```

and default file objects are not checked either
```
$ cwltool  --validate tests/wf/default-dir5.cwl
INFO /home/michael/cwltool/env3.10/bin/cwltool 3.1.20220902151754
INFO Resolved 'tests/wf/default-dir5.cwl' to 'file:///home/michael/cwltool/tests/wf/default-dir5.cwl'
tests/wf/default-dir5.cwl:20:13: Warning: Field `location` contains undefined reference to
                                 `file:///home/michael/cwltool/tests/wf/inp1`
tests/wf/default-dir5.cwl:20:13: Warning: Field `location` contains undefined reference to
                                 `file:///home/michael/cwltool/tests/wf/inp1`
tests/wf/default-dir5.cwl is valid CWL.
$ cwltool --fast-validator --validate tests/wf/default-dir5.cwl
INFO /home/michael/cwltool/env3.10/bin/cwltool 3.1.20220902151754
INFO Resolved 'tests/wf/default-dir5.cwl' to 'file:///home/michael/cwltool/tests/wf/default-dir5.cwl'
tests/wf/default-dir5.cwl is valid CWL.
```

And extensions don't work either. We should error on combining `--fast-validator` and `--enable-ext`
```
$ cwltool  --enable-ext --validate tests/wf/generator/pytoolgen.cwl
INFO /home/michael/cwltool/env3.10/bin/cwltool 3.1.20220902151754
INFO Resolved 'tests/wf/generator/pytoolgen.cwl' to 'file:///home/michael/cwltool/tests/wf/generator/pytoolgen.cwl'
tests/wf/generator/pytoolgen.cwl is valid CWL.

$ cwltool --fast-validator --enable-ext --validate tests/wf/generator/pytoolgen.cwl
INFO /home/michael/cwltool/env3.10/bin/cwltool 3.1.20220902151754
INFO Resolved 'tests/wf/generator/pytoolgen.cwl' to 'file:///home/michael/cwltool/tests/wf/generator/pytoolgen.cwl'
ERROR Tool definition failed validation:
- tried CommandLineTool but
  Not a CommandLineTool
- tried ExpressionTool but
  Not a ExpressionTool
- tried Workflow but
  Not a Workflow
- tried Operation but
  Not a Operation
- tried array<CommandLineTool | ExpressionTool | Workflow | Operation> but
  Expected a list, was <class 'dict'>
```
It is also intolerant of extra `id` fields, like in https://www.commonwl.org/v1.2/CommandLineTool.html#CommandInputRecordSchema found in `tests/wf/schemadef-bug-1473.cwl`

**Comment by mr-c** [comment]
Here's a mixed version situation that triggered an error (upgrading the reference document to CWL v1.2 made it go away):

(we should re-test when CWL v1.0 is supported on the fast path)

```
$ cwltool  --fast-validator tests/subgraph/1432.cwl --step1_in foo --step2_in bar
INFO /home/michael/cwltool/env3.10/bin/cwltool 3.1.20220902151754
INFO Resolved 'tests/subgraph/1432.cwl' to 'file:///home/michael/cwltool/tests/subgraph/1432.cwl'
ERROR I'm sorry, I couldn't load this CWL file, try again with --debug for more information.
The error was: 'LineCol' object has no attribute 'filename'
```
Execution succeeds without `--fast-validator`

**Comment by mr-c** [comment]
When `--validate` is given a `$graph` file with no `#main` or `main` entry, it validates all entries.

But with `--fast-validator` that fails:

```
$ cwltool  --validate tests/wf/packed_no_main.cwl
INFO /home/michael/cwltool/env3.10/bin/cwltool 3.1.20220902151754
INFO Resolved 'tests/wf/packed_no_main.cwl' to 'file:///home/michael/cwltool/tests/wf/packed_no_main.cwl'
WARNING:root:File contains $graph of multiple objects and no default process (#main). Validating all objects:
file:///home/michael/cwltool/tests/wf/packed_no_main.cwl#echo is valid CWL.
file:///home/michael/cwltool/tests/wf/packed_no_main.cwl#cat is valid CWL.
file:///home/michael/cwltool/tests/wf/packed_no_main.cwl#collision is valid CWL.
tests/wf/packed_no_main.cwl is valid CWL.

$ cwltool --fast-validator --validate tests/wf/packed_no_main.cwl
INFO /home/michael/cwltool/env3.10/bin/cwltool 3.1.20220902151754
INFO Resolved 'tests/wf/packed_no_main.cwl' to 'file:///home/michael/cwltool/tests/wf/packed_no_main.cwl'
WARNING:root:File contains $graph of multiple objects and no default process (#main). Validating all objects:
ERROR Tool definition failed validation:
Expected URI string, MutableMapping or MutableSequence, got <class 'NoneType'>
ERROR:cwltool:Tool definition failed validation:
Expected URI string, MutableMapping or MutableSequence, got <class 'NoneType'>
```

File: README.rst [hunk 1]
```diff
@@ -667,6 +667,19 @@ given in the following table; all are optional.
 +----------------+------------------+----------+------------------------------+
 
 
+Enabling Fast Parser (experimental)
+-----------------------------------
+
+For very large workflows, it can spend a lot of time in
```
Comment (mr-c) [inline README.rst:673]:
```suggestion
For very large workflows, `cwltool` can spend a lot of time in
```

**Review by mr-c — CHANGES_REQUESTED** [review]
odd that we are missing test coverage at https://github.com/common-workflow-language/cwltool/pull/1720/files#diff-7bc82a2b4dbfbc386a1c1d22dac00140d782b959e858cd99978fc3ee193d15f2R273 and beyond ; can you look into that? Is test coverage not being collected in the --fast-parser github action run of the conformance tests?

**Comment by tetron** [comment]
I don't actually know how we enable coverage data to be collected?

**Comment by mr-c** [comment]
> I don't actually know how we enable coverage data to be collected?

https://github.com/common-workflow-language/cwltool/blob/1e150ab6ad9ebd329186d31ad010352e65605ceb/conformance-test.sh#L86

**Comment by tetron** [comment]
Well now it's failing on "pinging codecov"  :facepalm:

**Comment by mr-c** [comment]
> Well now it's failing on "pinging codecov" facepalm

Do you have a link? I'm not seeing that..

**Comment by tetron** [comment]
> Do you have a link? I'm not seeing that..

It was one of the tests, but it seems to have just been a transient failure.

**Comment by mr-c** [comment]
Coverage is being recorded for "cwltool/cwltool/*.py` and `cwltool/*.py` https://app.codecov.io/gh/common-workflow-language/cwltool/pull/1720 ; this is probably a source of confusion

**Comment by mr-c** [comment]
> The diff coverage is 91.01%.

Much better! (it was previously 33%)

**Review by mr-c — APPROVED** [review]

**Pull request merged**